cmd/cluster-operator/capi/capi.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// code in the package copied from: https://github.com/kubernetes-sigs/cluster-api/blob/v1.2.5/main.go
+package capi
+
+import (
+	"context"
+	"fmt"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+	addonscontrollers "sigs.k8s.io/cluster-api/exp/addons/controllers"
+	expv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	expcontrollers "sigs.k8s.io/cluster-api/exp/controllers"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/webhooks"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"kurator.dev/kurator/cmd/cluster-operator/config"
+)
+
+var (
+	log = ctrl.Log.WithName("capi")
+)
+
+func InitControllers(ctx context.Context, opts *config.Options, mgr ctrl.Manager) error {
+	if err := setupReconcilers(ctx, opts, mgr); err != nil {
+		log.Error(err, "init capi reconcilers failed")
+		return fmt.Errorf("init capi reconciler failed: %w", err)
+	}
+
+	if err := setupWebhooks(mgr); err != nil {
+		log.Error(err, "init capi webhooks failed")
+		return fmt.Errorf("init capi webhooks failed: %w", err)
+	}
+
+	return nil
+}
+
+func setupReconcilers(ctx context.Context, opts *config.Options, mgr ctrl.Manager) error {
+	// Set up a ClusterCacheTracker and ClusterCacheReconciler to provide to controllers
+	// requiring a connection to a remote cluster
+	log := ctrl.Log.WithName("remote").WithName("ClusterCacheTracker")
+	tracker, err := remote.NewClusterCacheTracker(
```

Comment:
what does tracker do

Reply:
`ClusterCacheTracker.GetClient return client for remote cluster`

Reply:
I mean what is the responsibility of the controller

Reply:
update comment